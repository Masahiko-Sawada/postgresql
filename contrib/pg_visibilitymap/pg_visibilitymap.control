# pg_visibilitymap extension
comment = 'examine the visibility map (VM)'
default_version = '1.0'
module_pathname = '$libdir/pg_visibilitymap'
relocatable = true
