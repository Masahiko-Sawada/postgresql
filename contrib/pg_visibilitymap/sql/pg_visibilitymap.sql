CREATE EXTENSION pg_visibilitymap;

CREATE TABLE t1 AS SELECT * FROM generate_series(1,1000) AS p;
\d t1
VACUUM t1;
SELECT relpages FROM pg_class WHERE relname = 't1';
SELECT * FROM pg_visibilitymap('t1');
SELECT * FROM pg_visibility('t1');
DELETE FROM t1 WHERE p < 500;
SELECT * FROM pg_visibilitymap('t1');
SELECT * FROM pg_visibility('t1');
VACUUM FREEZE t1;
SELECT * FROM pg_visibilitymap('t1');
SELECT * FROM pg_visibility('t1');
DROP TABLE t1;

--
-- Errors
--
CREATE TABLE t2 (col int primary key);
INSERT INTO t2 SELECT generate_series(1,100);
CREATE VIEW t2_view AS (SELECT * FROM t2);
CREATE MATERIALIZED VIEW t2_matview AS (SELECT * FROM t2);
SELECT pg_visibilitymap('t2', 0);
SELECT pg_visibilitymap('t2_matview', 0);
SELECT pg_visibilitymap('t2_view', 0); -- error
SELECT pg_visibilitymap('t2_pkey', 0); -- error
SELECT pg_page_flags('t2', 0);
SELECT pg_page_flags('t2_matview', 0);
SELECT pg_page_flags('t2_view', 0); -- error
SELECT pg_page_flags('t2_pkey', 0); -- error
DROP TABLE t2 CASCADE;
