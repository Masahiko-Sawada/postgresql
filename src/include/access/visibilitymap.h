/*-------------------------------------------------------------------------
 *
 * visibilitymap.h
 *		visibility map interface
 *
 *
 * Portions Copyright (c) 2007-2015, PostgreSQL Global Development Group
 * Portions Copyright (c) 1994, Regents of the University of California
 *
 * src/include/access/visibilitymap.h
 *
 *-------------------------------------------------------------------------
 */
#ifndef VISIBILITYMAP_H
#define VISIBILITYMAP_H

#include "access/xlogdefs.h"
#include "storage/block.h"
#include "storage/buf.h"
#include "utils/relcache.h"

/* Flags for bit map */
#define VISIBILITYMAP_ALL_VISIBLE	0x01
#define VISIBILITYMAP_ALL_FROZEN	0x02

extern void visibilitymap_clear(Relation rel, BlockNumber heapBlk, Buffer vmbuf);
extern void visibilitymap_pin(Relation rel, BlockNumber heapBlk,
				  Buffer *vmbuf);
extern bool visibilitymap_pin_ok(BlockNumber heapBlk, Buffer vmbuf);
extern void visibilitymap_set(Relation rel, BlockNumber heapBlk, Buffer heapBuf,
							  XLogRecPtr recptr, Buffer vmBuf, TransactionId cutoff_xid,
							  uint8 flags);
extern bool visibilitymap_test(Relation rel, BlockNumber heapBlk, Buffer *vmbuf,
							   uint8 flags);
extern void visibilitymap_count(Relation rel, BlockNumber *all_visible,
									   BlockNumber *all_frozen);
extern void visibilitymap_truncate(Relation rel, BlockNumber nheapblocks);

#endif   /* VISIBILITYMAP_H */
