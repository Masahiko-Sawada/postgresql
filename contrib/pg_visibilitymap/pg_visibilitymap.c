/*-------------------------------------------------------------------------
 *
 * pg_visibilitymap.c
 *	  display contents of a visibility map and page level bits
 *
 *	  contrib/pg_visibilitymap/pg_visibilitymap.c
 *-------------------------------------------------------------------------
 */
#include "postgres.h"

#include "access/visibilitymap.h"
#include "funcapi.h"
#include "miscadmin.h"
#include "storage/bufmgr.h"
#include "utils/rel.h"


PG_MODULE_MAGIC;

PG_FUNCTION_INFO_V1(pg_visibilitymap);
PG_FUNCTION_INFO_V1(pg_page_flags);

Datum
pg_visibilitymap(PG_FUNCTION_ARGS)
{
	Oid			relid = PG_GETARG_OID(0);
	int64		blkno = PG_GETARG_INT64(1);
	int32		mapbits;
	Relation	rel;
	Buffer		vmbuffer = InvalidBuffer;

	if (!superuser())
		ereport(ERROR,
				(errcode(ERRCODE_INSUFFICIENT_PRIVILEGE),
				 (errmsg("must be superuser to use pg_visibilitymap functions"))));

	rel = relation_open(relid, AccessShareLock);

	if (rel->rd_rel->relkind != RELKIND_RELATION &&
		rel->rd_rel->relkind != RELKIND_MATVIEW)
		ereport(ERROR,
				(errcode(ERRCODE_WRONG_OBJECT_TYPE),
				 errmsg("\"%s\" is not a table or materialized view",
						RelationGetRelationName(rel))));

	if (blkno < 0 || blkno > MaxBlockNumber)
		ereport(ERROR,
				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
				 errmsg("invalid block number")));

	mapbits = (int32) visibilitymap_get_status(rel, blkno, &vmbuffer);
	if (vmbuffer != InvalidBuffer)
		ReleaseBuffer(vmbuffer);

	relation_close(rel, AccessShareLock);
	PG_RETURN_INT32(mapbits);
}

Datum
pg_page_flags(PG_FUNCTION_ARGS)
{
	Oid			relid = PG_GETARG_OID(0);
	int64		blkno = PG_GETARG_INT64(1);
	int32		pagebits;
	Relation	rel;
	Buffer		buffer;
	Page		page;

	if (!superuser())
		ereport(ERROR,
				(errcode(ERRCODE_INSUFFICIENT_PRIVILEGE),
				 (errmsg("must be superuser to use pg_visibilitymap functions"))));

	rel = relation_open(relid, AccessShareLock);

	if (rel->rd_rel->relkind != RELKIND_RELATION &&
		rel->rd_rel->relkind != RELKIND_MATVIEW)
		ereport(ERROR,
				(errcode(ERRCODE_WRONG_OBJECT_TYPE),
				 errmsg("\"%s\" is not a table or materialized view",
						RelationGetRelationName(rel))));

	if (blkno < 0 || blkno > MaxBlockNumber)
		ereport(ERROR,
				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
				 errmsg("invalid block number")));

	buffer = ReadBuffer(rel, blkno);
	LockBuffer(buffer, BUFFER_LOCK_SHARE);

	page = BufferGetPage(buffer);
	pagebits = (int32) (((PageHeader) (page))->pd_flags);

	UnlockReleaseBuffer(buffer);
	relation_close(rel, AccessShareLock);
	PG_RETURN_INT32(pagebits);
}
