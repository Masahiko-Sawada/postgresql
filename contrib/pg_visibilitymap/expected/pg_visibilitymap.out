CREATE EXTENSION pg_visibilitymap;
CREATE TABLE t1 AS SELECT * FROM generate_series(1,1000) AS p;
\d t1
      Table "public.t1"
 Column |  Type   | Modifiers 
--------+---------+-----------
 p      | integer | 

VACUUM t1;
SELECT relpages FROM pg_class WHERE relname = 't1';
 relpages 
----------
        5
(1 row)

SELECT * FROM pg_visibilitymap('t1');
 blkno | mapbits 
-------+---------
     0 |       1
     1 |       1
     2 |       1
     3 |       1
     4 |       1
(5 rows)

SELECT * FROM pg_visibility('t1');
 blkno | mapbits | pagebits 
-------+---------+----------
     0 |       1 |        4
     1 |       1 |        4
     2 |       1 |        4
     3 |       1 |        4
     4 |       1 |        4
(5 rows)

DELETE FROM t1 WHERE p < 500;
SELECT * FROM pg_visibilitymap('t1');
 blkno | mapbits 
-------+---------
     0 |       0
     1 |       0
     2 |       0
     3 |       1
     4 |       1
(5 rows)

SELECT * FROM pg_visibility('t1');
 blkno | mapbits | pagebits 
-------+---------+----------
     0 |       0 |        0
     1 |       0 |        0
     2 |       0 |        0
     3 |       1 |        4
     4 |       1 |        4
(5 rows)

VACUUM FREEZE t1;
SELECT * FROM pg_visibilitymap('t1');
 blkno | mapbits 
-------+---------
     0 |       3
     1 |       3
     2 |       3
     3 |       3
     4 |       3
(5 rows)

SELECT * FROM pg_visibility('t1');
 blkno | mapbits | pagebits 
-------+---------+----------
     0 |       3 |       13
     1 |       3 |       13
     2 |       3 |       13
     3 |       3 |       12
     4 |       3 |       12
(5 rows)

DROP TABLE t1;
--
-- Errors
--
CREATE TABLE t2 (col int primary key);
INSERT INTO t2 SELECT generate_series(1,100);
CREATE VIEW t2_view AS (SELECT * FROM t2);
CREATE MATERIALIZED VIEW t2_matview AS (SELECT * FROM t2);
SELECT pg_visibilitymap('t2', 0);
 pg_visibilitymap 
------------------
                0
(1 row)

SELECT pg_visibilitymap('t2_matview', 0);
 pg_visibilitymap 
------------------
                0
(1 row)

SELECT pg_visibilitymap('t2_view', 0); -- error
ERROR:  "t2_view" is not a table or materialized view
SELECT pg_visibilitymap('t2_pkey', 0); -- error
ERROR:  "t2_pkey" is not a table or materialized view
SELECT pg_page_flags('t2', 0);
 pg_page_flags 
---------------
             0
(1 row)

SELECT pg_page_flags('t2_matview', 0);
 pg_page_flags 
---------------
             0
(1 row)

SELECT pg_page_flags('t2_view', 0); -- error
ERROR:  "t2_view" is not a table or materialized view
SELECT pg_page_flags('t2_pkey', 0); -- error
ERROR:  "t2_pkey" is not a table or materialized view
DROP TABLE t2 CASCADE;
NOTICE:  drop cascades to 2 other objects
DETAIL:  drop cascades to view t2_view
drop cascades to materialized view t2_matview
